### PR TITLE
Split compiler and runtime build systems

### PR DESCRIPTION
**Pull request by meow464**
Let's see if it passes

### CODE REVIEW

**Comment by meow464** [comment]
@certik Ready for review. This only splits the build system but cross compilation still works. To build only the runtime simply call cmake with the src/runtime dir instead of the project's root dir.

**Comment by certik** [comment]
This looks good, thanks! I tested it, and this is the behavior I observed:
```console
(lf) lfortran/lfortran(cross-compilation-runtime) $ ./build0.sh 
+ ci/version.sh
++ git describe --tags --dirty
+ version=v0.18.0-559-gc5f6dec00
+ version=0.18.0-559-gc5f6dec00
+ echo 0.18.0-559-gc5f6dec00
+ python src/libasr/asdl_cpp.py grammar/AST.asdl src/lfortran/ast.h
+ python src/libasr/asdl_cpp.py src/libasr/ASR.asdl src/libasr/asr.h
+ python src/libasr/wasm_instructions_visitor.py
Assuming default values of wasm_instructions.txt and wasm_visitor.h
+ cd src/lfortran/parser
+ re2c -W -b tokenizer.re -o tokenizer.cpp
+ cd src/lfortran/parser
+ re2c -W -b preprocessor.re -o preprocessor.cpp
+ cd src/lfortran/parser
+ bison -Wall -d -r all parser.yy
./+ grep -n ''\''' src/lfortran/parser/parser.yy
+ echo OK
OK
(lf) lfortran/lfortran(cross-compilation-runtime) $ ./build1.sh 
++ pwd
+ cmake -DCMAKE_BUILD_TYPE=Debug -DWITH_LLVM=yes -DLFORTRAN_BUILD_ALL=yes -DWITH_STACKTRACE=yes '-DCMAKE_PREFIX_PATH=;/Users/ondrej/mambaforge/envs/lf' -DCMAKE_INSTALL_PREFIX=/Users/ondrej/repos/lfortran/lfortran/inst .
+ ci/version.sh
++ git describe --tags --dirty
+ version=v0.18.0-559-gc5f6dec00
+ version=0.18.0-559-gc5f6dec00
+ echo 0.18.0-559-gc5f6dec00
+ python src/libasr/asdl_cpp.py grammar/AST.asdl src/lfortran/ast.h
+ python src/libasr/asdl_cpp.py src/libasr/ASR.asdl src/libasr/asr.h
+ python src/libasr/wasm_instructions_visitor.py
Assuming default values of wasm_instructions.txt and wasm_visitor.h
+ cd src/lfortran/parser
+ re2c -W -b tokenizer.re -o tokenizer.cpp
+ cd src/lfortran/parser
+ re2c -W -b preprocessor.re -o preprocessor.cpp
+ cd src/lfortran/parser
+ bison -Wall -d -r all parser.yy
+ grep -n ''\''' src/lfortran/parser/parser.yy
OK
+ echo OK
-- The C compiler identification is AppleClang 13.0.0.13000029
-- The CXX compiler identification is AppleClang 13.0.0.13000029
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /Library/Developer/CommandLineTools/usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /Library/Developer/CommandLineTools/usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found ZLIB: /Users/ondrej/mambaforge/envs/lf/lib/libz.a (found version "1.2.12") 
-- Found LLVM 11.1.0
-- Using LLVMConfig.cmake in: /Users/ondrej/mambaforge/envs/lf/lib/cmake/llvm
-- Found MACHO: /Library/Developer/CommandLineTools/SDKs/MacOSX12.1.sdk/usr/include  


Configuration results
---------------------
LFORTRAN_VERSION: 0.18.0-559-gc5f6dec00
CPACK_PACKAGE_FILE_NAME: lfortran-0.18.0-559-gc5f6dec00-Darwin
C compiler      : /Library/Developer/CommandLineTools/usr/bin/cc
C++ compiler    : /Library/Developer/CommandLineTools/usr/bin/c++
Build type: Debug
C compiler flags      : -g
C++ compiler flags    : -Wall -Wextra -g -ggdb
Installation prefix: /Users/ondrej/repos/lfortran/lfortran/inst
WITH_LFORTRAN_ASSERT: yes
LFORTRAN_STATIC_BIN: no
LFORTRAN_BUILD_TO_WASM: no
WITH_STACKTRACE: yes
WITH_UNWIND: yes
WITH_BFD: no
WITH_DWARFDUMP: yes
WITH_LINKH: no
WITH_MACHO: yes
HAVE_LFORTRAN_DEMANGLE: yes
WITH_LLVM: yes
WITH_XEUS: no
WITH_JSON: no
WITH_FMT: no
WITH_LFORTRAN_BINARY_MODFILES: YES
WITH_RUNTIME_LIBRARY: YES
WITH_WHEREAMI: yes
WITH_ZLIB: yes
WITH_TARGET_AARCH64: yes
WITH_TARGET_X86: no
WITH_TARGET_WASM: no
-- Configuring done
-- Generating done
-- Build files have been written to: /Users/ondrej/repos/lfortran/lfortran
+ cmake --build . -j16 --target install
[  2%] Building C object src/runtime/legacy/CMakeFiles/lfortran_runtime.dir/__/__/libasr/runtime/lfortran_intrinsics.c.o
[  3%] Building CXX object src/tests/CMakeFiles/doctest.dir/doctest.cpp.o
[  3%] Building CXX object src/libasr/CMakeFiles/asr.dir/codegen/x86_assembler.cpp.o
[  4%] Building C object src/runtime/legacy/CMakeFiles/lfortran_runtime_static.dir/__/__/libasr/runtime/lfortran_intrinsics.c.o
[  5%] Building CXX object src/libasr/CMakeFiles/asr.dir/codegen/asr_to_julia.cpp.o
[  6%] Building CXX object src/libasr/CMakeFiles/asr.dir/codegen/asr_to_c.cpp.o
[  8%] Building CXX object src/libasr/CMakeFiles/asr.dir/codegen/wasm_to_x64.cpp.o
[  8%] Building CXX object src/libasr/CMakeFiles/asr.dir/codegen/asr_to_cpp.cpp.o
[  9%] Building CXX object src/libasr/CMakeFiles/asr.dir/codegen/asr_to_py.cpp.o
[ 10%] Building CXX object src/libasr/CMakeFiles/asr.dir/pass/param_to_const.cpp.o
[ 11%] Building CXX object src/libasr/CMakeFiles/asr.dir/pass/do_loops.cpp.o
[ 12%] Building CXX object src/libasr/CMakeFiles/asr.dir/codegen/wasm_to_wat.cpp.o
[ 13%] Building CXX object src/libasr/CMakeFiles/asr.dir/codegen/asr_to_wasm.cpp.o
[ 14%] Building CXX object src/libasr/CMakeFiles/asr.dir/codegen/asr_to_x86.cpp.o
[ 15%] Building CXX object src/libasr/CMakeFiles/asr.dir/codegen/wasm_utils.cpp.o
[ 16%] Building CXX object src/libasr/CMakeFiles/asr.dir/codegen/wasm_to_x86.cpp.o
[ 17%] Linking C static library liblfortran_runtime_static.a
[ 18%] Linking C shared library ../liblfortran_runtime.dylib
[ 18%] Built target lfortran_runtime_static
[ 19%] Building CXX object src/libasr/CMakeFiles/asr.dir/pass/for_all.cpp.o
[ 19%] Built target lfortran_runtime
[ 20%] Building CXX object src/libasr/CMakeFiles/asr.dir/pass/global_stmts.cpp.o
[ 21%] Building CXX object src/libasr/CMakeFiles/asr.dir/pass/global_stmts_program.cpp.o
[ 22%] Building CXX object src/libasr/CMakeFiles/asr.dir/pass/select_case.cpp.o
[ 23%] Building CXX object src/libasr/CMakeFiles/asr.dir/pass/implied_do_loops.cpp.o
[ 24%] Building CXX object src/libasr/CMakeFiles/asr.dir/pass/array_op.cpp.o
[ 25%] Building CXX object src/libasr/CMakeFiles/asr.dir/pass/subroutine_from_function.cpp.o
[ 26%] Building CXX object src/libasr/CMakeFiles/asr.dir/pass/class_constructor.cpp.o
[ 27%] Building CXX object src/libasr/CMakeFiles/asr.dir/pass/arr_slice.cpp.o
[ 28%] Building CXX object src/libasr/CMakeFiles/asr.dir/pass/print_arr.cpp.o
[ 29%] Building CXX object src/libasr/CMakeFiles/asr.dir/pass/print_list.cpp.o
[ 30%] Building CXX object src/libasr/CMakeFiles/asr.dir/pass/pass_utils.cpp.o
[ 31%] Building CXX object src/libasr/CMakeFiles/asr.dir/pass/unused_functions.cpp.o
[ 32%] Building CXX object src/libasr/CMakeFiles/asr.dir/pass/flip_sign.cpp.o
[ 33%] Building CXX object src/libasr/CMakeFiles/asr.dir/pass/div_to_mul.cpp.o
[ 34%] Linking CXX static library libdoctest.a
[ 34%] Built target doctest
[ 35%] Building CXX object src/libasr/CMakeFiles/asr.dir/pass/fma.cpp.o
[ 36%] Building CXX object src/libasr/CMakeFiles/asr.dir/pass/loop_vectorise.cpp.o
[ 37%] Building CXX object src/libasr/CMakeFiles/asr.dir/pass/sign_from_value.cpp.o
[ 38%] Building CXX object src/libasr/CMakeFiles/asr.dir/pass/inline_function_calls.cpp.o
[ 39%] Building CXX object src/libasr/CMakeFiles/asr.dir/pass/dead_code_removal.cpp.o
[ 40%] Building CXX object src/libasr/CMakeFiles/asr.dir/pass/loop_unroll.cpp.o
[ 41%] Building CXX object src/libasr/CMakeFiles/asr.dir/pass/instantiate_template.cpp.o
[ 42%] Building CXX object src/libasr/CMakeFiles/asr.dir/pass/update_array_dim_intrinsic_calls.cpp.o
[ 43%] Building CXX object src/libasr/CMakeFiles/asr.dir/pass/pass_array_by_data.cpp.o
[ 44%] Building CXX object src/libasr/CMakeFiles/asr.dir/pass/pass_list_expr.cpp.o
[ 45%] Building CXX object src/libasr/CMakeFiles/asr.dir/pass/pass_compare.cpp.o
[ 46%] Building CXX object src/libasr/CMakeFiles/asr.dir/asr_verify.cpp.o
[ 47%] Building CXX object src/libasr/CMakeFiles/asr.dir/asr_utils.cpp.o
[ 48%] Building CXX object src/libasr/CMakeFiles/asr.dir/diagnostics.cpp.o
[ 49%] Building CXX object src/libasr/CMakeFiles/asr.dir/stacktrace.cpp.o
[ 50%] Building CXX object src/libasr/CMakeFiles/asr.dir/string_utils.cpp.o
[ 51%] Building CXX object src/libasr/CMakeFiles/asr.dir/asr_scopes.cpp.o
[ 52%] Building CXX object src/libasr/CMakeFiles/asr.dir/modfile.cpp.o
[ 53%] Building CXX object src/libasr/CMakeFiles/asr.dir/serialization.cpp.o
[ 54%] Building CXX object src/libasr/CMakeFiles/asr.dir/utils2.cpp.o
[ 55%] Building CXX object src/libasr/CMakeFiles/asr.dir/codegen/evaluator.cpp.o
[ 56%] Building CXX object src/libasr/CMakeFiles/asr.dir/codegen/asr_to_llvm.cpp.o
[ 57%] Building CXX object src/libasr/CMakeFiles/asr.dir/codegen/llvm_array_utils.cpp.o
[ 58%] Building CXX object src/libasr/CMakeFiles/asr.dir/codegen/llvm_utils.cpp.o
[ 59%] Building CXX object src/libasr/CMakeFiles/asr.dir/pass/nested_vars.cpp.o
[ 60%] Linking CXX static library libasr.a
[ 60%] Built target asr
[ 62%] Building CXX object src/lfortran/CMakeFiles/lfortran_lib.dir/parser/tokenizer.cpp.o
[ 62%] Building CXX object src/lfortran/CMakeFiles/lfortran_lib.dir/parser/parser.tab.cc.o
[ 63%] Building CXX object src/lfortran/CMakeFiles/lfortran_lib.dir/parser/preprocessor.cpp.o
[ 64%] Building CXX object src/lfortran/CMakeFiles/lfortran_lib.dir/semantics/ast_to_asr.cpp.o
[ 65%] Building CXX object src/lfortran/CMakeFiles/lfortran_lib.dir/semantics/ast_symboltable_visitor.cpp.o
[ 66%] Building CXX object src/lfortran/CMakeFiles/lfortran_lib.dir/parser/parser.cpp.o
[ 67%] Building CXX object src/lfortran/CMakeFiles/lfortran_lib.dir/cwrapper.cpp.o
[ 68%] Building CXX object src/lfortran/CMakeFiles/lfortran_lib.dir/parser/fixedform_tokenizer.cpp.o
[ 69%] Building CXX object src/lfortran/CMakeFiles/lfortran_lib.dir/pickle.cpp.o
[ 70%] Building CXX object src/lfortran/CMakeFiles/lfortran_lib.dir/fortran_evaluator.cpp.o
[ 72%] Building CXX object src/lfortran/CMakeFiles/lfortran_lib.dir/semantics/ast_body_visitor.cpp.o
[ 72%] Building CXX object src/lfortran/CMakeFiles/lfortran_lib.dir/ast_serialization.cpp.o
[ 73%] Building CXX object src/lfortran/CMakeFiles/lfortran_lib.dir/ast_to_src.cpp.o
[ 74%] Building CXX object src/lfortran/CMakeFiles/lfortran_lib.dir/ast_to_openmp.cpp.o
[ 75%] Building CXX object src/lfortran/CMakeFiles/lfortran_lib.dir/utils.cpp.o
[ 76%] Building CXX object src/lfortran/CMakeFiles/lfortran_lib.dir/mod_to_asr.cpp.o
[ 77%] Building CXX object src/lfortran/CMakeFiles/lfortran_lib.dir/__/bin/tpl/whereami/whereami.cpp.o
[ 78%] Linking CXX static library liblfortran_lib.a
[ 78%] Built target lfortran_lib
[ 80%] Building CXX object src/bin/CMakeFiles/lfortran.dir/lfortran.cpp.o
[ 82%] Building C object src/lfortran/tests/CMakeFiles/test_cwrapper.dir/test_cwrapper.c.o
[ 82%] Building CXX object src/lfortran/tests/CMakeFiles/test_stacktrace.dir/test_stacktrace.cpp.o
[ 82%] Building CXX object src/bin/CMakeFiles/parse.dir/parse.cpp.o
[ 83%] Building CXX object src/bin/CMakeFiles/parse2.dir/parse2.cpp.o
[ 84%] Building CXX object src/lfortran/tests/CMakeFiles/test_lfortran.dir/test_parse.cpp.o
[ 85%] Building CXX object src/lfortran/tests/CMakeFiles/test_lfortran.dir/test_asm.cpp.o
[ 86%] Building CXX object src/bin/CMakeFiles/cpptranslate.dir/cpptranslate.cpp.o
[ 88%] Building CXX object src/lfortran/tests/CMakeFiles/test_lfortran.dir/test_serialization.cpp.o
[ 88%] Building CXX object src/lfortran/tests/CMakeFiles/test_lfortran.dir/test_error_rendering.cpp.o
[ 89%] Building CXX object src/lfortran/tests/CMakeFiles/test_lfortran.dir/test_llvm.cpp.o
[ 90%] Building CXX object src/lfortran/tests/CMakeFiles/test_lfortran.dir/test_stacktrace2.cpp.o
[ 91%] Building CXX object src/lfortran/tests/CMakeFiles/test_lfortran.dir/test_ast.cpp.o
[ 92%] Building CXX object src/lfortran/tests/CMakeFiles/test_lfortran.dir/test_pickle.cpp.o
[ 93%] Linking CXX executable test_cwrapper
[ 93%] Built target test_cwrapper
[ 94%] Linking CXX executable test_stacktrace
[ 94%] Built target test_stacktrace
[ 95%] Linking CXX executable parse
[ 96%] Linking CXX executable parse2
[ 96%] Built target parse
[ 96%] Built target parse2
[ 97%] Linking CXX executable test_lfortran
[ 97%] Built target test_lfortran
[ 98%] Linking CXX executable cpptranslate
[ 98%] Built target cpptranslate
[100%] Linking CXX executable lfortran
[100%] Built target lfortran
Install the project...
-- Install configuration: "Debug"
-- Installing: /Users/ondrej/repos/lfortran/lfortran/inst/bin/lfortran
-- Installing: /Users/ondrej/repos/lfortran/lfortran/inst/share/lfortran/lib/liblfortran_runtime_static.a
-- Installing: /Users/ondrej/repos/lfortran/lfortran/inst/share/lfortran/lib/liblfortran_runtime.dylib
-- Installing: /Users/ondrej/repos/lfortran/lfortran/inst/share/lfortran/lib/impure/lfortran_intrinsics.h
'/Users/ondrej/repos/spack/opt/spack/darwin-monterey-m1/apple-clang-13.0.0/cmake-3.23.1-3geuicgdlylivqqft4pjwrdnmb63dfxr/bin/cmake' '-G Unix Makefiles' '-DCMAKE_TOOLCHAIN_FILE=' '-DCMAKE_INSTALL_PREFIX=/Users/ondrej/repos/lfortran/lfortran/inst' '-DCMAKE_BUILD_TYPE=Debug' '-DCMAKE_PREFIX_PATH=;/Users/ondrej/mambaforge/envs/lf' '-DCMAKE_VERBOSE_MAKEFILE:BOOL=FALSE' '-DCMAKE_Fortran_COMPILER=/Users/ondrej/repos/lfortran/lfortran/inst/bin/lfortran' '/Users/ondrej/repos/lfortran/lfortran/src/runtime/'
-- The C compiler identification is AppleClang 13.0.0.13000029
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /Library/Developer/CommandLineTools/usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Configuring done
-- Generating done
-- Build files have been written to: /Users/ondrej/repos/lfortran/lfortran/src/runtime
-- Done.
'/Users/ondrej/repos/spack/opt/spack/darwin-monterey-m1/apple-clang-13.0.0/cmake-3.23.1-3geuicgdlylivqqft4pjwrdnmb63dfxr/bin/cmake' '--build' '.' '--target' 'install'
Scanning dependencies of target lfortran_intrinsic_builtin
Scanning dependencies of target lfortran_intrinsic_kind
Scanning dependencies of target lfortran_intrinsic_iso_fortran_env
[  5%] Building Fortran object CMakeFiles/lfortran_intrinsic_builtin.dir/builtin/lfortran_intrinsic_builtin.f90.o
[ 11%] Building C object legacy/CMakeFiles/lfortran_runtime_static.dir/Users/ondrej/repos/lfortran/lfortran/src/libasr/runtime/lfortran_intrinsics.c.o
[ 23%] Building Fortran object CMakeFiles/lfortran_intrinsic_kind.dir/pure/lfortran_intrinsic_kind.f90.o
[ 23%] Building C object legacy/CMakeFiles/lfortran_runtime.dir/Users/ondrej/repos/lfortran/lfortran/src/libasr/runtime/lfortran_intrinsics.c.o
[ 29%] Building Fortran object CMakeFiles/lfortran_intrinsic_iso_fortran_env.dir/pure/lfortran_intrinsic_iso_fortran_env.f90.o
[ 41%] Linking C shared library ../liblfortran_runtime.dylib
[ 41%] Linking C static library liblfortran_runtime_static.a
[ 41%] Built target lfortran_intrinsic_iso_fortran_env
[ 41%] Built target lfortran_intrinsic_kind
[ 41%] Built target lfortran_intrinsic_builtin
[ 41%] Built target lfortran_runtime_static
Scanning dependencies of target lfortran_intrinsic_trig
Scanning dependencies of target lfortran_intrinsic_ieee_arithmetic
Scanning dependencies of target lfortran_intrinsic_optimization
Scanning dependencies of target lfortran_intrinsic_bit
[ 47%] Building Fortran object CMakeFiles/lfortran_intrinsic_trig.dir/pure/lfortran_intrinsic_trig.f90.o
[ 52%] Building Fortran object CMakeFiles/lfortran_intrinsic_optimization.dir/builtin/lfortran_intrinsic_optimization.f90.o
[ 58%] Building Fortran object CMakeFiles/lfortran_intrinsic_ieee_arithmetic.dir/pure/lfortran_intrinsic_ieee_arithmetic.f90.o
Scanning dependencies of target lfortran_intrinsic_iso_c_binding
[ 64%] Building Fortran object CMakeFiles/lfortran_intrinsic_bit.dir/impure/lfortran_intrinsic_bit.f90.o
Scanning dependencies of target lfortran_intrinsic_math3
Scanning dependencies of target lfortran_intrinsic_string
[ 64%] Built target lfortran_runtime
[ 70%] Building Fortran object CMakeFiles/lfortran_intrinsic_math3.dir/pure/lfortran_intrinsic_math3.f90.o
[ 76%] Building Fortran object CMakeFiles/lfortran_intrinsic_string.dir/pure/lfortran_intrinsic_string.f90.o
[ 82%] Building Fortran object CMakeFiles/lfortran_intrinsic_iso_c_binding.dir/pure/lfortran_intrinsic_iso_c_binding.f90.o
[ 82%] Built target lfortran_intrinsic_optimization
[ 82%] Built target lfortran_intrinsic_ieee_arithmetic
[ 82%] Built target lfortran_intrinsic_trig
[ 82%] Built target lfortran_intrinsic_bit
[ 82%] Built target lfortran_intrinsic_math3
[ 82%] Built target lfortran_intrinsic_iso_c_binding
[ 82%] Built target lfortran_intrinsic_string
Scanning dependencies of target lfortran_intrinsic_math2
[ 88%] Building Fortran object CMakeFiles/lfortran_intrinsic_math2.dir/pure/lfortran_intrinsic_math2.f90.o
[ 88%] Built target lfortran_intrinsic_math2
Scanning dependencies of target lfortran_intrinsic_math
[ 94%] Building Fortran object CMakeFiles/lfortran_intrinsic_math.dir/impure/lfortran_intrinsic_math.f90.o
[ 94%] Built target lfortran_intrinsic_math
Scanning dependencies of target lfortran_intrinsic_sin
[100%] Building Fortran object CMakeFiles/lfortran_intrinsic_sin.dir/impure/lfortran_intrinsic_sin.f90.o
[100%] Built target lfortran_intrinsic_sin
Install the project...
-- Install configuration: "Debug"
-- Installing: /Users/ondrej/repos/lfortran/lfortran/inst/share/lfortran/lib/lfortran_intrinsic_builtin.mod
-- Installing: /Users/ondrej/repos/lfortran/lfortran/inst/share/lfortran/lib/lfortran_intrinsic_optimization.mod
-- Installing: /Users/ondrej/repos/lfortran/lfortran/inst/share/lfortran/lib/lfortran_intrinsic_kind.mod
-- Installing: /Users/ondrej/repos/lfortran/lfortran/inst/share/lfortran/lib/lfortran_intrinsic_iso_fortran_env.mod
-- Installing: /Users/ondrej/repos/lfortran/lfortran/inst/share/lfortran/lib/lfortran_intrinsic_ieee_arithmetic.mod
-- Installing: /Users/ondrej/repos/lfortran/lfortran/inst/share/lfortran/lib/lfortran_intrinsic_iso_c_binding.mod
-- Installing: /Users/ondrej/repos/lfortran/lfortran/inst/share/lfortran/lib/lfortran_intrinsic_math2.mod
-- Installing: /Users/ondrej/repos/lfortran/lfortran/inst/share/lfortran/lib/lfortran_intrinsic_math3.mod
-- Installing: /Users/ondrej/repos/lfortran/lfortran/inst/share/lfortran/lib/lfortran_intrinsic_math.mod
-- Installing: /Users/ondrej/repos/lfortran/lfortran/inst/share/lfortran/lib/lfortran_intrinsic_trig.mod
-- Installing: /Users/ondrej/repos/lfortran/lfortran/inst/share/lfortran/lib/lfortran_intrinsic_sin.mod
-- Installing: /Users/ondrej/repos/lfortran/lfortran/inst/share/lfortran/lib/lfortran_intrinsic_string.mod
-- Installing: /Users/ondrej/repos/lfortran/lfortran/inst/share/lfortran/lib/lfortran_intrinsic_bit.mod
-- Up-to-date: /Users/ondrej/repos/lfortran/lfortran/inst/share/lfortran/lib/liblfortran_runtime_static.a
-- Installing: /Users/ondrej/repos/lfortran/lfortran/inst/share/lfortran/lib/liblfortran_runtime.dylib
-- Up-to-date: /Users/ondrej/repos/lfortran/lfortran/inst/share/lfortran/lib/impure/lfortran_intrinsics.h
-- Done.
(lf) lfortran/lfortran(cross-compilation-runtime) $ make
Consolidate compiler generated dependencies of target asr
[ 54%] Built target asr
Consolidate compiler generated dependencies of target doctest
[ 56%] Built target doctest
src/runtime/legacy/CMakeFiles/lfortran_runtime_static.dir/build.make:62: legacy/CMakeFiles/lfortran_runtime_static.dir/depend.make: No such file or directory
src/runtime/legacy/CMakeFiles/lfortran_runtime_static.dir/build.make:64: legacy/CMakeFiles/lfortran_runtime_static.dir/compiler_depend.make: No such file or directory
src/runtime/legacy/CMakeFiles/lfortran_runtime_static.dir/build.make:67: legacy/CMakeFiles/lfortran_runtime_static.dir/progress.make: No such file or directory
src/runtime/legacy/CMakeFiles/lfortran_runtime_static.dir/build.make:70: legacy/CMakeFiles/lfortran_runtime_static.dir/flags.make: No such file or directory
make[2]: *** No rule to make target `legacy/CMakeFiles/lfortran_runtime_static.dir/flags.make'.  Stop.
make[1]: *** [src/runtime/legacy/CMakeFiles/lfortran_runtime_static.dir/all] Error 2
make: *** [all] Error 2
```
So the second time you run "make", you get an error message. 

If you modify a file and rebuild:
```console
(lf) lfortran/lfortran(cross-compilation-runtime) $ touch src/libasr/diagnostics.cpp 
(lf) lfortran/lfortran(cross-compilation-runtime) $ make
[  1%] Building CXX object src/libasr/CMakeFiles/asr.dir/diagnostics.cpp.o
[  2%] Linking CXX static library libasr.a
[ 54%] Built target asr
[ 56%] Built target doctest
src/runtime/legacy/CMakeFiles/lfortran_runtime_static.dir/build.make:62: legacy/CMakeFiles/lfortran_runtime_static.dir/depend.make: No such file or directory
src/runtime/legacy/CMakeFiles/lfortran_runtime_static.dir/build.make:64: legacy/CMakeFiles/lfortran_runtime_static.dir/compiler_depend.make: No such file or directory
src/runtime/legacy/CMakeFiles/lfortran_runtime_static.dir/build.make:67: legacy/CMakeFiles/lfortran_runtime_static.dir/progress.make: No such file or directory
src/runtime/legacy/CMakeFiles/lfortran_runtime_static.dir/build.make:70: legacy/CMakeFiles/lfortran_runtime_static.dir/flags.make: No such file or directory
make[2]: *** No rule to make target `legacy/CMakeFiles/lfortran_runtime_static.dir/flags.make'.  Stop.
make[1]: *** [src/runtime/legacy/CMakeFiles/lfortran_runtime_static.dir/all] Error 2
make: *** [all] Error 2
```

It's probably related to how the runtime library build is integrated with the main cmake.

**Comment by certik** [comment]
What are your requirements? You need to build the runtime library using cmake+lfortran as a standalone cmake project?

**Comment by meow464** [comment]
> You need to build the runtime library using cmake+lfortran as a standalone cmake project? 

Yes, cmake wants to find the compiler before building the project. Even if we tell it to not test the compiler it still wants to find it.

I will test it some more and see if I figure a way around this.

**Comment by certik** [comment]
Yes, if we want to use cmake to drive LFortran, then indeed the lfortran binary has to be ready before cmake is invoked, so we have to have a separate cmake project for the runtime library.

All we need is to just figure out how to integrate the two cmake projects, so that if you modify a file, everything gets correctly rebuilt.

**Comment by meow464** [comment]
@certik Should be fixed now. The problem was that make likes to `cd` into directories and run a command, ninja likes to use full paths. The last time I had a problem like this it was the other way around, make worked and ninja didn't.

**Comment by meow464** [comment]
The runtime can only be built after install (after the compiler is there), to rebuild everything again you need `make install` or `ninja install`. To rebuild just the runtime `cd src/runtime` and run make or ninja.

**Comment by certik** [comment]
I will test by hand first before merging.

**Comment by certik** [comment]
I tested it and I think it starts to look very good, touching LFortran source files correctly rebuilds only what is needed.

However, this doesn't seem to work:
```
$ touch src/runtime/pure/lfortran_intrinsic_ieee_arithmetic.f90
$ make
[ 55%] Built target asr
[ 57%] Built target doctest
[ 59%] Built target lfortran_runtime_static
[ 77%] Built target lfortran_lib
[ 79%] Built target test_cwrapper
[ 81%] Built target test_stacktrace
[ 90%] Built target test_lfortran
[ 92%] Built target lfortran
[ 94%] Built target parse
[ 96%] Built target parse2
[ 98%] Built target cpptranslate
[100%] Built target lfortran_runtime
```
How can one rebuild the runtime library?

**Comment by meow464** [comment]
Yes, to build only the runtime you need to enter the runtime directory. To rebuild everything from the top directory the command would be `make install`. The runtime can only be built after the compiler has been installed, i.e. after the install target.

I think the runtime could be built with just `make` (before the install target) if we could guarantee lfortran always works from the build directory. Currently I think this translates to a commitment to using relative paths for the linking command and `liblfortran_runtime.so` and `liblfortran_runtime_static.a` and possibly other things. Might be better to wait until we are no longer in beta.

**Comment by certik** [comment]
Yes, LFortran always works from the build directory. I see, so there is another problem. I did this diff:
```diff
--- a/build1.sh
+++ b/build1.sh
@@ -9,6 +9,5 @@ cmake \
     -DLFORTRAN_BUILD_ALL=yes \
     -DWITH_STACKTRACE=yes \
     -DCMAKE_PREFIX_PATH="$CMAKE_PREFIX_PATH_LFORTRAN;$CONDA_PREFIX" \
-    -DCMAKE_INSTALL_PREFIX=`pwd`/inst \
     .
-cmake --build . -j16 --target install
+make -j16
```
Which we should still support. The way I work is that we build it in the build directory, but do not install. If I run it, I get:
```
[ 99%] Linking CXX executable cpptranslate
[ 99%] Built target cpptranslate
[100%] Linking CXX executable lfortran
[100%] Built target lfortran
$ cd src/runtime 
$ make
make: *** No targets specified and no makefile found.  Stop.
```
In other words, if you don't install LFortran, you cannot build the runtime library it seems, and it does not build automatically.

I would not tie this to the installed location, rather build this using the compiled LFortran in the build directory.

**Comment by czgdp1807** [comment]
So seems like some more work needs to be done here. Converting to draft. Please re-mark it ready for review if the case is otherwise.

**Comment by meow464** [comment]
Everything should be in order now. `make build` rebuilds everything, `make install` installs everything.
One question: When building to wasm the [second call to cmake](https://github.com/lfortran/lfortran/blob/main/build_to_wasm.sh#L19) with emcmake doesn't build the runtime right? I don't get mod files on src/runtime when testing locally but I'm asking just to be totally sure.

**Review by czgdp1807 — APPROVED** [review]

**Comment by certik** [comment]
I tested this by hand, it looks like everything works. This is about the same speed:
```
$ touch src/libasr/utils2.cpp
$ time make lfortran
```

I noticed just doing "make" got about 2x slower (from 0.5s to 1s). But `make -j` is still pretty fast, so I think that's fine. I will now review the change itself.

File: src/runtime/CMakeLists.txt [hunk 1]
```diff
@@ -1,19 +1,96 @@
...
+lfrotran_compile_runtime(lfortran_intrinsic_trig pure)
+add_dependencies(lfortran_intrinsic_trig lfortran_intrinsic_iso_fortran_env)
+
+lfrotran_compile_runtime(lfortran_intrinsic_math impure)
+add_dependencies(lfortran_intrinsic_math lfortran_intrinsic_iso_c_binding)
+add_dependencies(lfortran_intrinsic_math lfortran_intrinsic_builtin)
+add_dependencies(lfortran_intrinsic_math lfortran_intrinsic_iso_fortran_env)
+add_dependencies(lfortran_intrinsic_math lfortran_intrinsic_math2)
+
+lfrotran_compile_runtime(lfortran_intrinsic_sin impure)
```
Comment (certik) [inline src/runtime/CMakeLists.txt:65]:
```suggestion
lfortran_compile_runtime(lfortran_intrinsic_sin impure)
```

Comment (meow464) [reply src/runtime/CMakeLists.txt]: Thanks for spotting it!

**Comment by meow464** [comment]
> I tested this by hand, it looks like everything works. This is about the same speed:
> 
> ```
> $ touch src/libasr/utils2.cpp
> $ time make lfortran
> ```
> 
> I noticed just doing "make" got about 2x slower (from 0.5s to 1s). But `make -j` is still pretty fast, so I think that's fine. I will now review the change itself.

This might be from the additional call to cmake and subsequent configure step.

**Review by certik — APPROVED** [review]
Looks like tests are passing. I think this looks good now, thanks for this work @meow464 and sorry it took so long to finish. This cleans it up nicely and should help with cross compilation as well.

**Comment by meow464** [comment]
>sorry it took so long to finish

I'm glad we were able to make it perfect.

>This cleans it up nicely

Thanks. I also have been looking at the pull request to integrate build0.sh and friends into cmake and the issue to compile lapack with cmake and lfortran.

**Comment by certik** [comment]
Perfect! Yes, just open up issues for all bugs in LFortran for Lapack. Some files compile, but some don't yet.

Yes, the `build0.sh` should be integrated, that would be very helpful. It's also very tricky to get it done right, but it might be possible.

**Comment by certik** [comment]
@meow464 do you think you can please have a look at this regression? We need to fix it.

@yharaskrik42 go ahead and create a new issue for this.

**Pull request merged**